<table>
  <tr>
    <td><p style="font-size:45px; color: #55BBD2">Analysis of light microscopy images in Python</p></td>
    <td><img src="../ressources/lmb_logo.svg" alt="LMB Logo" width="500" height="600" align="right"></td>
  </tr>
</table>
<table>
  <tr>
    <td><p style="font-size:15px; color: #55BBD2">Version: September 2025</p></td>
  </tr>
</table>

# Part 4: Particle tracking

<b>Problem:</b> Track transcription factor CDX2

<b>Dataset:</b> 181228_CDX2_9s_c48_n073.tif

<b>Credit:</b> Kuhn, Timo; Hettich, Johannes; Davtyan, Rubina; Gebhardt, J. Christof M. (2021), Single molecule tracking and analysis framework including theory-predicted parameter settings, Scientific Reports, Journal-article, [https://doi.org/10.1038/s41598-021-88802-7](https://doi.org/10.1038/s41598-021-88802-7). [Link to dataset](https://datadryad.org/dataset/doi:10.5061/dryad.0zpc866wh).

<b>Objectives:</b>
- Detect particles (skimage.feature.blob_dog)
- Join dataframes (pandas.concat, pandas.merge)
- Track particles by linking their positions over time (laptrack.LapTrack)
- Visualize tracks and 3D rendering using napari (napari.add_tracks)

<b>Workflow:</b>

<img src="../ressources/workflow/workflow3.png" alt="drawing" width="800"/>

In [1]:
from pathlib import Path
from bioio import BioImage

data_folder = Path('../data')
particles = BioImage(Path(data_folder, '181228_CDX2_9s_c48_n073.tif'))
particles_data = particles.data

sz = particles_data.shape
print(sz)

(105, 1, 1, 114, 168)


## Napari display
Now we are going to display the particles images, consisting of 105 frames, using the napari plugin. To be able to visualize the 3D rendering of the data, we need to ensure we are displaying a 3D image in one layer. In other words, we need to specify the channel and z = 0 in the viewer i.e display particles_data[:,0,0].

In [2]:
import napari

v = napari.Viewer()
v.add_image(particles_data[:,0,0])

<Image layer 'Image' at 0x12c7fa450>

## Detect each particle (spot) using blob_dog from the feature module

The blob_dog is another function for spot detection. It is a faster approximation of the LoG approach and has few parameters that we need to care for in order to filter out what we consider particles and which ones are not. There are importants parameters that we will cover here: threshold, min_sigma and max_sigma. 

- threshold: The absolute lower bound for scale space maxima. Local maxima smaller than threshold are ignored. Reduce this to detect blobs with lower intensities. <br>
- min_sigma: The minimum standard deviation for Gaussian kernel.
- max_sigma: The maximum standard deviation for Gaussian kernel.

Further reading: https://scikit-image.org/docs/stable/api/skimage.feature.html#skimage.feature.blob_dog

We will save the positions of the detected particles in a panda dataframe.

In [3]:
from skimage import feature 
import pandas as pd

particles_positions = []

for t in range(sz[0]):    
    coordinates = feature.blob_dog(particles_data[t].squeeze(), min_sigma=1, max_sigma=5, threshold=0.005)    
    particles_positions.append(pd.DataFrame({'x': coordinates[:,1], 'y': coordinates[:,0], 'w': coordinates[:,2], 'frame':t}))

particles_positions = pd.concat(particles_positions)
    
particles_positions

x     y     w  frame
0    64.0  30.0  1.00      0
1   104.0  15.0  1.00      0
2   114.0  37.0  1.00      0
3    76.0  34.0  1.00      0
4   108.0  95.0  1.00      0
..    ...   ...   ...    ...
10   36.0  65.0  1.60    104
11  103.0  22.0  1.00    104
12   36.0  20.0  1.60    104
13   74.0  11.0  2.56    104
14   76.0  97.0  1.00    104

[2061 rows x 4 columns]

We do not have tracks yet here but points. We can use add_points to add the points to the napari viewer. We will need to select the dataframe from the the TYX information, where T is the timepoints or frames.

<div class="alert alert-success">
       
#### Exercise 

Add the points to the napari viewer using particles_positions[['frame', 'y', 'x']]. Set the size to 1, opacity to 0.5, and face_color to 'green'.
    
</div>

<div class="alert alert-success">

#### Exercise       

The results we have here were optimized using threshold = 0.005, min_sigma = 1, and max_sigma = 5. You may find a better setting than these. <br> 
So go back to the spot detection cell and play with the values of those three parameters and observe how the results are changing accordingly.

   
</div>

## Particles trajectories
We are now going to trace the trajectories of the particles by linking the coordinates positions. <br> 
For this, we will use a Python package named 'laptrack'.

In [ ]:
from laptrack import LapTrack

lt = LapTrack(cutoff=5**2)
track_df, _, _ = lt.predict_dataframe(particles_positions, ["y", "x"], only_coordinate_cols=False)
track_df = track_df.reset_index()

v.add_tracks(track_df[["track_id", "frame", "y", "x"]], tail_length=50)

<Tracks layer 'Tracks [3]' at 0x148eb8b90>

<div class="alert alert-success">

#### Exercise       

Laptrack has a hands-on list of parameters that we can tune to improve the tracking results. The essential parameters are listed below. Play with those to understand how it works.
   
- "dist_metric": the global metric to use to calculate any distance. Default value is "sqeuclidean". With this, any input distance should be squared distance. For "euclidean", the distance is used not its square.

- "cutoff" or "track_cost_cutoff": maximum allowed distance between particles to form a track.

- "gap_closing_cost_cutoff": cost cutoff for gap closing i.e linking one track to another. It is the maximum allowed distance between last frame of one track and the start of the another track to link the two tracks as one.

- "gap_closing_max_frame_count": maximum frame gaps

Further reading: [here](https://laptrack.readthedocs.io/en/stable/reference.html)
</div>

## Additional constraints

But maybe we eventually want to remove particles that are only appearing in few frames?

Let say we won't consider the trajectories of particles that are appearing in less than 10 frames.

In [5]:
import pandas as pd

l = track_df['track_id'].unique() 
allowed_track_length = 10
particles_id = [k for k in l if len(track_df.loc[track_df['track_id']==k])>=allowed_track_length]

selected_tracks = pd.DataFrame([])

for t in range(len(particles_id)):
    coordinates = track_df.loc[track_df['track_id']==particles_id[t]]
    selected_tracks = pd.concat([selected_tracks, coordinates])

selected_tracks

frame  index     x     y    w  frame_y  tree_id  track_id
0         0      0  64.0  30.0  1.0        0        0         0
17        1      2  64.0  30.0  1.0        1        0         0
57        3     11  62.0  42.0  1.0        3        0         0
68        4      3  68.0  43.0  1.0        4        0         0
118       6     13  62.0  38.0  1.0        6        0         0
...     ...    ...   ...   ...  ...      ...      ...       ...
1989    100      5  87.0  93.0  1.0      100      229       229
2004    101      3  87.0  93.0  1.0      101      229       229
2018    102      1  88.0  93.0  1.0      102      229       229
2034    103      2  88.0  93.0  1.0      103      229       229
2049    104      3  90.0  86.0  1.0      104      229       229

[1529 rows x 8 columns]

<div class="alert alert-success">

#### Exercise       

Change the allowed_track_length and observe the number of trajectories displayed in the overlay display.
   
</div>

## Exporting tabular results

To save results in a csv file use the export function from pandas. <br>
The excelsheet will be save in your current directory. To check where is that directory, run the command pwd in a new cell and you will see the csv file be saved in that directory.


In [6]:
pwd

'/Users/ratsimandresy/Documents/GitHub/introduction-to-python-bioimage-analysis/nbs'

In [7]:
selected_tracks.to_csv('tracking_results.csv')

# Some more helpful tips

- scientific images are data, they can be compromised by inappropriate manipulations
- take good images:
    - don't oversaturate your data
    - use the full dynamic range when taking your images
- do not segment on the data you are measuring, use a housekeeping channel
- images that are compared to each other need to be processed and acquired in the same manner